src/ray/object_manager/object_manager.cc

```diff
@@ -1,156 +1,309 @@
 };
 
 ray::Status ObjectManager::Pull(const ObjectID &object_id, const ClientID &client_id) {
+  // Check if object is already local, and client_id is not itself.
+  if (local_objects_.count(object_id) != 0 || client_id == client_id_) {
+    return ray::Status::OK();
+  }
+
   Status status =
-      GetMsgConnection(client_id, [this, object_id](SenderConnection::pointer client) {
-        Status status = ExecutePull(object_id, client);
-      });
+      connection_pool_.GetSender(ConnectionPool::MESSAGE, client_id,
+                                 [this, object_id](SenderConnection::pointer conn) {
+                                   Status status = ExecutePull(object_id, conn);
+                                 },
+                                 [this, object_id]() {
+                                   // connection failed, so reschedule pull.
+                                   SchedulePull(object_id, config_.pull_timeout_ms);
+                                 });
   return status;
 };
 
 ray::Status ObjectManager::ExecutePull(const ObjectID &object_id,
                                        SenderConnection::pointer conn) {
-  size_t message_type = OMMessageType_PullRequest;
-  boost::system::error_code error_code;
-  boost::asio::write(conn->GetSocket(),
-                     boost::asio::buffer(&message_type, sizeof(message_type)),
-                     error_code);
   flatbuffers::FlatBufferBuilder fbb;
-  auto message = CreatePullRequest(fbb, fbb.CreateString(client_id_.binary()),
-                                   fbb.CreateString(object_id.binary()));
+  auto message = CreatePullRequestMessage(fbb, fbb.CreateString(client_id_.binary()),
+                                          fbb.CreateString(object_id.binary()));
   fbb.Finish(message);
-  size_t length = fbb.GetSize();
-  std::vector<boost::asio::const_buffer> buffer;
-  buffer.push_back(boost::asio::buffer(&length, sizeof(length)));
-  buffer.push_back(boost::asio::buffer(fbb.GetBufferPointer(), length));
-  boost::asio::write(conn->GetSocket(), buffer);
+  (void)conn->WriteMessage(OMMessageType_PullRequest, fbb.GetSize(),
+                           fbb.GetBufferPointer());
+  (void)connection_pool_.ReleaseSender(ConnectionPool::MESSAGE, conn);
   return ray::Status::OK();
 };
 
 ray::Status ObjectManager::Push(const ObjectID &object_id, const ClientID &client_id) {
+  main_service_->post(
+      [this, object_id, client_id]() { RAY_CHECK_OK(Push_(object_id, client_id)); });
+  return Status::OK();
+};
+
+ray::Status ObjectManager::Push_(const ObjectID &object_id, const ClientID &client_id) {
+  //  RAY_LOG(INFO) << "Push_ "
+  //                << object_id << " "
+  //                << client_id << " ";
+  transfer_queue_.QueueSend(client_id, object_id);
+  return DequeueTransfers();
+}
+
+ray::Status ObjectManager::DequeueTransfers() {
+  ray::Status status = ray::Status::OK();
+  while (num_transfers_ < max_transfers_) {
```

Comment:
Is there ever a case where a thread can pop more than one transfer from the queue? If not, then I would just make this an `if`.

Reply:
This isn't necessarily invoked by the thread that executes the transfer. The method initiates the execution of a transfer by invoking `ExecuteSend` or `ExecuteReceive`.